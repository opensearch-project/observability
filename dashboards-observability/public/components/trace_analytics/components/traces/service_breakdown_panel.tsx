/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiFlexGroup,
  EuiFlexItem,
  EuiHealth,
  EuiHorizontalRule,
  EuiPanel,
  EuiSpacer,
  EuiText,
} from '@elastic/eui';
import _ from 'lodash';
import React, { useMemo } from 'react';
import { Plt } from '../../../visualizations/plotly/plot';
import { PanelTitle } from '../common/helper_functions';

export function ServiceBreakdownPanel(props: { data: Plotly.Data[] }) {
  const layout = useMemo(
    () =>
      ({
        height: 200,
        width: 200,
        showlegend: false,
        margin: {
          l: 5,
          r: 5,
          b: 5,
          t: 5,
        },
      } as Partial<Plotly.Layout>),
    [props.data]
  );

  const renderStats = () => {
    return props.data.length > 0 ? (
      <EuiFlexGroup responsive={false} style={{ maxHeight: 260, overflowY: 'auto' }}>
        <EuiFlexItem>
          <EuiFlexGroup direction="column" alignItems="flexStart" gutterSize="m" responsive={false}>
            {props.data[0].marker.colors.map((color, i) => (
              <EuiFlexItem key={`label-${i}`}>
                <EuiHealth color={color}>
                  <div style={{ whiteSpace: 'nowrap' }}>{props.data[0].labels[i]}</div>
                </EuiHealth>
              </EuiFlexItem>
            ))}
          </EuiFlexGroup>
        </EuiFlexItem>
        <EuiFlexItem />
        <EuiFlexItem />
        <EuiFlexItem>
          <EuiFlexGroup direction="column" alignItems="flexEnd" gutterSize="m" responsive={false}>
            {props.data[0].values.map((value, i) => (
              <EuiFlexItem key={`value-${i}`}>
                <EuiText size="s">{_.round(value, 2)}%</EuiText>
              </EuiFlexItem>
            ))}
          </EuiFlexGroup>
        </EuiFlexItem>
      </EuiFlexGroup>
    ) : null;
  };

  const stats = useMemo(() => renderStats(), [props.data]);

  return (
    <>
      <EuiPanel>
        <PanelTitle title="Time spent by service" />
        <EuiHorizontalRule margin="m" />
        <EuiFlexGroup direction="column" alignItems="center">
          <EuiFlexItem>
            {props.data?.length > 0 ? <Plt data={props.data} layout={layout} /> : null}
          </EuiFlexItem>
          <EuiSpacer />
          <EuiFlexItem>{stats}</EuiFlexItem>
        </EuiFlexGroup>
        <EuiSpacer />
      </EuiPanel>
    </>
  );
}
