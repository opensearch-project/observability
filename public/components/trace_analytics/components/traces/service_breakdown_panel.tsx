/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 *   Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import {
  EuiFlexGroup,
  EuiFlexItem,
  EuiHealth,
  EuiHorizontalRule,
  EuiPanel,
  EuiSpacer,
  EuiText,
} from '@elastic/eui';
import _ from 'lodash';
import React, { useMemo } from 'react';
import { Plt } from '../../../visualizations/plotly/plot';
import { PanelTitle } from '../common/helper_functions';

export function ServiceBreakdownPanel(props: { data: Plotly.Data[] }) {
  const layout = useMemo(
    () =>
      ({
        height: 200,
        width: 200,
        showlegend: false,
        margin: {
          l: 5,
          r: 5,
          b: 5,
          t: 5,
        },
      } as Partial<Plotly.Layout>),
    [props.data]
  );

  const renderStats = () => {
    return props.data.length > 0 ? (
      <EuiFlexGroup responsive={false} style={{ maxHeight: 260, overflowY: 'auto' }}>
        <EuiFlexItem>
          <EuiFlexGroup direction="column" alignItems="flexStart" gutterSize="m" responsive={false}>
            {props.data[0].marker.colors.map((color, i) => (
              <EuiFlexItem key={`label-${i}`}>
                <EuiHealth color={color}>
                  <div style={{ whiteSpace: 'nowrap' }}>{props.data[0].labels[i]}</div>
                </EuiHealth>
              </EuiFlexItem>
            ))}
          </EuiFlexGroup>
        </EuiFlexItem>
        <EuiFlexItem />
        <EuiFlexItem />
        <EuiFlexItem>
          <EuiFlexGroup direction="column" alignItems="flexEnd" gutterSize="m" responsive={false}>
            {props.data[0].values.map((value, i) => (
              <EuiFlexItem key={`value-${i}`}>
                <EuiText size="s">{_.round(value, 2)}%</EuiText>
              </EuiFlexItem>
            ))}
          </EuiFlexGroup>
        </EuiFlexItem>
      </EuiFlexGroup>
    ) : null;
  };

  const stats = useMemo(() => renderStats(), [props.data]);

  return (
    <>
      <EuiPanel>
        <PanelTitle title="Time spent by service" />
        <EuiHorizontalRule margin="m" />
        <EuiFlexGroup direction="column" alignItems="center">
          <EuiFlexItem>
            {props.data?.length > 0 ? <Plt data={props.data} layout={layout} /> : null}
          </EuiFlexItem>
          <EuiSpacer />
          <EuiFlexItem>{stats}</EuiFlexItem>
        </EuiFlexGroup>
        <EuiSpacer />
      </EuiPanel>
    </>
  );
}
