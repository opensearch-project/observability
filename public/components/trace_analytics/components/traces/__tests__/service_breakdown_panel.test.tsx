/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { ServiceBreakdownPanel } from '../service_breakdown_panel';

describe('Service breakdown panel component', () => {
  configure({ adapter: new Adapter() });

  it('renders empty service breakdown panel', () => {
    const wrapper = mount(<ServiceBreakdownPanel data={[]} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders service breakdown panel', () => {
    const data = [
      {
        values: [100],
        labels: ['inventory'],
        marker: { colors: ['#7492e7'] },
        type: 'pie',
        textinfo: 'none',
        hovertemplate: '%{label}<br>%{value:.2f}%<extra></extra>',
      },
    ] as Plotly.Data[];
    const wrapper = mount(<ServiceBreakdownPanel data={data} />);
    expect(wrapper).toMatchSnapshot();
  });
});
