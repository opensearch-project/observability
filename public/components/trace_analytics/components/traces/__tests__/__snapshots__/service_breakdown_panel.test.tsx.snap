// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Service breakdown panel component renders empty service breakdown panel 1`] = `
<ServiceBreakdownPanel
  data={Array []}
>
  <EuiPanel>
    <div
      className="euiPanel euiPanel--paddingMedium"
    >
      <PanelTitle
        title="Time spent by service"
      >
        <EuiText
          size="m"
        >
          <div
            className="euiText euiText--medium"
          >
            <span
              className="panel-title"
            >
              Time spent by service
            </span>
          </div>
        </EuiText>
      </PanelTitle>
      <EuiHorizontalRule
        margin="m"
      >
        <hr
          className="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginMedium"
        />
      </EuiHorizontalRule>
      <EuiFlexGroup
        alignItems="center"
        direction="column"
      >
        <div
          className="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--directionColumn euiFlexGroup--responsive"
        >
          <EuiFlexItem>
            <div
              className="euiFlexItem"
            />
          </EuiFlexItem>
          <EuiSpacer>
            <div
              className="euiSpacer euiSpacer--l"
            />
          </EuiSpacer>
          <EuiFlexItem>
            <div
              className="euiFlexItem"
            />
          </EuiFlexItem>
        </div>
      </EuiFlexGroup>
      <EuiSpacer>
        <div
          className="euiSpacer euiSpacer--l"
        />
      </EuiSpacer>
    </div>
  </EuiPanel>
</ServiceBreakdownPanel>
`;

exports[`Service breakdown panel component renders service breakdown panel 1`] = `
<ServiceBreakdownPanel
  data={
    Array [
      Object {
        "hovertemplate": "%{label}<br>%{value:.2f}%<extra></extra>",
        "labels": Array [
          "inventory",
        ],
        "marker": Object {
          "colors": Array [
            "#7492e7",
          ],
        },
        "textinfo": "none",
        "type": "pie",
        "values": Array [
          100,
        ],
      },
    ]
  }
>
  <EuiPanel>
    <div
      className="euiPanel euiPanel--paddingMedium"
    >
      <PanelTitle
        title="Time spent by service"
      >
        <EuiText
          size="m"
        >
          <div
            className="euiText euiText--medium"
          >
            <span
              className="panel-title"
            >
              Time spent by service
            </span>
          </div>
        </EuiText>
      </PanelTitle>
      <EuiHorizontalRule
        margin="m"
      >
        <hr
          className="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginMedium"
        />
      </EuiHorizontalRule>
      <EuiFlexGroup
        alignItems="center"
        direction="column"
      >
        <div
          className="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--directionColumn euiFlexGroup--responsive"
        >
          <EuiFlexItem>
            <div
              className="euiFlexItem"
            >
              <Plt
                data={
                  Array [
                    Object {
                      "hovertemplate": "%{label}<br>%{value:.2f}%<extra></extra>",
                      "labels": Array [
                        "inventory",
                      ],
                      "marker": Object {
                        "colors": Array [
                          "#7492e7",
                        ],
                      },
                      "textinfo": "none",
                      "type": "pie",
                      "values": Array [
                        100,
                      ],
                    },
                  ]
                }
                layout={
                  Object {
                    "height": 200,
                    "margin": Object {
                      "b": 5,
                      "l": 5,
                      "r": 5,
                      "t": 5,
                    },
                    "showlegend": false,
                    "width": 200,
                  }
                }
              >
                <PlotlyComponent
                  config={
                    Object {
                      "displayModeBar": false,
                    }
                  }
                  data={
                    Array [
                      Object {
                        "hovertemplate": "%{label}<br>%{value:.2f}%<extra></extra>",
                        "labels": Array [
                          "inventory",
                        ],
                        "marker": Object {
                          "colors": Array [
                            "#7492e7",
                          ],
                        },
                        "textinfo": "none",
                        "type": "pie",
                        "values": Array [
                          100,
                        ],
                      },
                    ]
                  }
                  debug={false}
                  layout={
                    Object {
                      "autosize": true,
                      "barmode": "stack",
                      "height": 200,
                      "hovermode": "closest",
                      "legend": Object {
                        "orientation": "h",
                        "traceorder": "normal",
                      },
                      "margin": Object {
                        "b": 5,
                        "l": 5,
                        "r": 5,
                        "t": 5,
                      },
                      "showlegend": false,
                      "width": 200,
                      "xaxis": Object {
                        "rangemode": "normal",
                        "showgrid": true,
                        "zeroline": false,
                      },
                      "yaxis": Object {
                        "rangemode": "normal",
                        "showgrid": true,
                        "zeroline": false,
                      },
                    }
                  }
                  style={
                    Object {
                      "height": "100%",
                      "width": "100%",
                    }
                  }
                  useResizeHandler={true}
                >
                  <div
                    style={
                      Object {
                        "height": "100%",
                        "width": "100%",
                      }
                    }
                  />
                </PlotlyComponent>
              </Plt>
            </div>
          </EuiFlexItem>
          <EuiSpacer>
            <div
              className="euiSpacer euiSpacer--l"
            />
          </EuiSpacer>
          <EuiFlexItem>
            <div
              className="euiFlexItem"
            >
              <EuiFlexGroup
                responsive={false}
                style={
                  Object {
                    "maxHeight": 260,
                    "overflowY": "auto",
                  }
                }
              >
                <div
                  className="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow"
                  style={
                    Object {
                      "maxHeight": 260,
                      "overflowY": "auto",
                    }
                  }
                >
                  <EuiFlexItem>
                    <div
                      className="euiFlexItem"
                    >
                      <EuiFlexGroup
                        alignItems="flexStart"
                        direction="column"
                        gutterSize="m"
                        responsive={false}
                      >
                        <div
                          className="euiFlexGroup euiFlexGroup--gutterMedium euiFlexGroup--alignItemsFlexStart euiFlexGroup--directionColumn"
                        >
                          <EuiFlexItem
                            key="label-0"
                          >
                            <div
                              className="euiFlexItem"
                            >
                              <EuiHealth
                                color="#7492e7"
                              >
                                <div
                                  className="euiHealth"
                                >
                                  <EuiFlexGroup
                                    alignItems="center"
                                    gutterSize="xs"
                                    responsive={false}
                                  >
                                    <div
                                      className="euiFlexGroup euiFlexGroup--gutterExtraSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow"
                                    >
                                      <EuiFlexItem
                                        grow={false}
                                      >
                                        <div
                                          className="euiFlexItem euiFlexItem--flexGrowZero"
                                        >
                                          <EuiIcon
                                            color="#7492e7"
                                            type="dot"
                                          >
                                            <EuiIconEmpty
                                              aria-hidden={true}
                                              className="euiIcon euiIcon--medium euiIcon-isLoading"
                                              focusable="false"
                                              role="img"
                                              style={
                                                Object {
                                                  "fill": "#7492e7",
                                                }
                                              }
                                            >
                                              <svg
                                                aria-hidden={true}
                                                className="euiIcon euiIcon--medium euiIcon-isLoading"
                                                focusable="false"
                                                height={16}
                                                role="img"
                                                style={
                                                  Object {
                                                    "fill": "#7492e7",
                                                  }
                                                }
                                                viewBox="0 0 16 16"
                                                width={16}
                                                xmlns="http://www.w3.org/2000/svg"
                                              />
                                            </EuiIconEmpty>
                                          </EuiIcon>
                                        </div>
                                      </EuiFlexItem>
                                      <EuiFlexItem
                                        grow={false}
                                      >
                                        <div
                                          className="euiFlexItem euiFlexItem--flexGrowZero"
                                        >
                                          <div
                                            style={
                                              Object {
                                                "whiteSpace": "nowrap",
                                              }
                                            }
                                          >
                                            inventory
                                          </div>
                                        </div>
                                      </EuiFlexItem>
                                    </div>
                                  </EuiFlexGroup>
                                </div>
                              </EuiHealth>
                            </div>
                          </EuiFlexItem>
                        </div>
                      </EuiFlexGroup>
                    </div>
                  </EuiFlexItem>
                  <EuiFlexItem>
                    <div
                      className="euiFlexItem"
                    />
                  </EuiFlexItem>
                  <EuiFlexItem>
                    <div
                      className="euiFlexItem"
                    />
                  </EuiFlexItem>
                  <EuiFlexItem>
                    <div
                      className="euiFlexItem"
                    >
                      <EuiFlexGroup
                        alignItems="flexEnd"
                        direction="column"
                        gutterSize="m"
                        responsive={false}
                      >
                        <div
                          className="euiFlexGroup euiFlexGroup--gutterMedium euiFlexGroup--alignItemsFlexEnd euiFlexGroup--directionColumn"
                        >
                          <EuiFlexItem
                            key="value-0"
                          >
                            <div
                              className="euiFlexItem"
                            >
                              <EuiText
                                size="s"
                              >
                                <div
                                  className="euiText euiText--small"
                                >
                                  100
                                  %
                                </div>
                              </EuiText>
                            </div>
                          </EuiFlexItem>
                        </div>
                      </EuiFlexGroup>
                    </div>
                  </EuiFlexItem>
                </div>
              </EuiFlexGroup>
            </div>
          </EuiFlexItem>
        </div>
      </EuiFlexGroup>
      <EuiSpacer>
        <div
          className="euiSpacer euiSpacer--l"
        />
      </EuiSpacer>
    </div>
  </EuiPanel>
</ServiceBreakdownPanel>
`;
